### ISSUE TITLE
Missing request body when handling a POST via superagent

### ISSUE THREAD

**Issue by Enet4** [mswjs/msw#278]
## Environment

| Name | Version |
| ---- | ------- |
| msw  | 0.19.5   |
| node | 12.18.2  |
| OS   | Windows 10 |

## Request handlers

```js
    // just replicates the request data back to the client
    rest.post('http://mock.example.net:8080/foo', (req, res, ctx) => {
        return res(
            ctx.status(200),
            ctx.body(req.body),
        )
    })
```

## Actual request

```js
const superagent = require('superagent')

let response = await superagent.post('http://mock.example.net:8080/foo')
    .send({
        bar: 6,
        baz: '34',
    })

expect(response.body).toMatchObject({
    bar: 6,
    baz: '34',
})
```

## Current behavior

Requests made via [superagent](https://visionmedia.github.io/superagent) are intercepted, but `req.body` at the respective response resolver always comes out as an empty string, regardless of the contents provided in the HTTP request. This was verified by peeking `req` in a debugger. Full Jest test below. The equivalent request using `http.request` directly does not reproduce the problem.

## Expected behavior

I would have expected `req.body` to contain the request's data in the HTTP message body (in the example, an object `{bar: 6, baz: '34'}`, although the behavior is not specific to this object nor to a JSON payload).
The request itself is well formed and behaves well in other contexts (example programs to a non-mocked server appear to work fine).

## Test

```javascript
const superagent = require('superagent')
const { rest } = require('msw')
const { setupServer } = require('msw/node')

const MOCKED_URL = 'http://mock.example.net:8080/foo'
let server = setupServer(
    rest.post(MOCKED_URL, (req, res, ctx) => {
        return res(
            ctx.status(200),
            ctx.body(req.body),
        )
    }),
)

beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

test('send POST request with body via superagent', async () => {
    let response = await superagent.post(MOCKED_URL)
        .send({
            bar: 6,
            baz: '34',
        })

    expect(response.body).toMatchObject({
        bar: 6,
        baz: '34',
    })
})
```

**Comment by kettanaito** [comment]
Hey, @Enet4. Thanks for reporting this.

We have tackled this issue as a part of #275 (fix in #277). You can try it yourself updating to `node-request-interceptor@0.3.2` (a dependency of `msw`). The fix will be released in the next `0.20.0` release.

**Comment by Enet4** [comment]
Thank you for the quick reply, @kettanaito. I could not yet validate this since I triggered another error, but maybe I am doing something wrong in the process of depending on the upstream version. I'm not sure if this is worth reporting, but I'll just be leaving a few details here. It should be OK for me to wait for an official release nevertheless.

The steps:

1. git clone this repository to a sibling directory
2. `yarn build`
3. in the test project, add `"msw": "file:../msw"` as a devDependency.
4. `yarn && yarn test`

```
    TypeError: The first argument must be of type string or an instance of Buffer, ArrayBuffer, or Array or an Array-like Object. Received an instance of ObjectTypeError [ERR_INVALID_ARG_TYPE]: The first argument must be of type string or an instance of Buffer, ArrayBuffer, or Array or an Array-like Object. Received an instance of Object

      at ClientRequestOverride.<anonymous> (node_modules/msw/node_modules/node-request-interceptor/lib/http/ClientRequest/ClientRequestOverride.js:210:62)
      at step (node_modules/msw/node_modules/node-request-interceptor/lib/http/ClientRequest/ClientRequestOverride.js:33:23)
      at Object.next (node_modules/msw/node_modules/node-request-interceptor/lib/http/ClientRequest/ClientRequestOverride.js:14:53)    
      at fulfilled (node_modules/msw/node_modules/node-request-interceptor/lib/http/ClientRequest/ClientRequestOverride.js:5:58)
```

**Comment by kettanaito** [comment]
That looks like you are sending a request body that is not string, nor Buffer. The error says you are giving an Object as your request body. Could you please double check that and make sure the request body is stringified? 

I can also recommend to check your usage with [this `supertest` regression test scenario](https://github.com/mswjs/node-request-interceptor/blob/master/test/regressions/http-supertest-post.test.ts). That's useful for both you checking the deviation points, and us making sure we are not missing any important points during testing.

**Comment by Enet4** [comment]
> That looks like you are sending a request body that is not string, nor Buffer. The error says you are giving an Object as your request body. Could you please double check that and make sure the request body is stringified?

That would be true, but the API on superagent intentionally allows you to send an object, which I expect to be serialized to JSON, as documented [here](https://visionmedia.github.io/superagent/#post-put-requests). Considering that the library works in practice, I am at a loss on what I could look into here. The supertest regression test also relies on this behaviour, on line 34.

The same test above still fails with the same error if I send a stringified JSON.

```javascript
test('send POST request with body via superagent', async () => {
    let response = await superagent.post(MOCKED_URL)
        .set('Content-Type', 'application/json')
        .send('{"bar":6,"baz":"34"}')

    expect(response.body).toMatchObject({
        bar: 6,
        baz: '34',
    })
})
```

I can indeed send a string as plain text (`.type('text')`), but this would not be desirable here.

> I can also recommend to check your usage with this supertest regression test scenario. That's useful for both you checking the deviation points, and us making sure we are not missing any important points during testing.

Well, this matter is intriguing me a bit more now. :) I wrote a new test adapted from this one which replaces supertest with superagent and serves the same Express.js application on the host's TCP port. It passes. What should I look into next?

```typescript
import express from 'express'
import bodyParser from 'body-parser'
import superagent from 'superagent'
import { RequestInterceptor, InterceptedRequest } from '../../src'
import { Server } from 'http'

let pool: InterceptedRequest[] = []
let interceptor: RequestInterceptor
let server: Server

const app = express()
app.use(bodyParser.json())
app.post('/', (req, res) => {
  res.status(200).json(req.body)
})

beforeAll(() => {
  interceptor = new RequestInterceptor()
  interceptor.use((req) => {
    pool.push(req)
  })
})

beforeEach((done) => {
  server = app.listen(8084, 'localhost', done)
})

afterEach(() => {
  pool = []
  server.close()
})

afterAll(() => {
  interceptor.restore()
})

test('preserves original POST request JSON body', async (done) => {
  const res = await superagent
    .post('http://localhost:8084/')
    .send({ query: 'foo' })

  expect(res.error).toBeFalsy()
  expect(res.status).toBe(200)
  expect(res.body).toEqual({ query: 'foo' })

  expect(pool).toHaveLength(1)
  expect(pool[0].method).toBe('POST')
  expect(pool[0].body).toEqual(JSON.stringify({ query: 'foo' }))

  done()
})
```

**Comment by kettanaito** [comment]
Could you please update to `msw@0.20.x` and let me know if the issue is still reproducible?
If so, please share a reproduction scenario with us (in your fork, for example) so we could take a look. Thanks.

**Comment by Enet4** [comment]
Thank you for the follow-up. One of the issues in the example above was that indeed I should not pass `req.body` to `ctx.body` if `req.body` isn't a string. I probably tried `ctx.body` instead of `ctx.json` because the latter was not giving me any properties. Starting from 0.20.0, using `ctx.json` would work as expected.

I eventually stumbled upon another issue related with the use of cookies in environments without `document` (#312), but it has since been resolved and released in 0.20.2 (#313).

Consider this resolved. :+1:
